crates/vhost/src/vhost_user/backend_req_handler.rs

```diff
@@ -519,6 +548,44 @@ impl<S: VhostUserBackendReqHandler> BackendReqHandler<S> {
                 let res = self.backend.remove_mem_region(&msg);
                 self.send_ack_message(&hdr, res)?;
             }
+            Ok(FrontendReq::SET_DEVICE_STATE_FD) => {
+                let file = take_single_file(files).ok_or(Error::IncorrectFds)?;
+                let msg =
+                    self.extract_request_body::<VhostUserTransferDeviceState>(&hdr, size, &buf)?;
+                let reply_hdr = self.new_reply_header::<VhostUserU64>(&hdr, 0)?;
+
+                let direction: VhostTransferStateDirection = msg
+                    .direction
+                    .try_into()
+                    .map_err(|_| Error::InvalidMessage)?;
+                let phase: VhostTransferStatePhase =
+                    msg.phase.try_into().map_err(|_| Error::InvalidMessage)?;
+                let res = self.backend.set_device_state_fd(direction, phase, file);
+
+                // The value returned is both an indication for success, and whether a file
+                // descriptor for a back-end-provided channel is returned: Bits 0–7 are 0 on
+                // success, and non-zero on error. Bit 8 is the invalid FD flag; this flag is
+                // set when there is no file descriptor returned.
+                match res {
+                    Ok(None) => {
+                        let msg = VhostUserU64::new(0x100); // set invalid FD flag
+                        self.main_sock.send_message(&reply_hdr, &msg, None)?;
+                    }
+                    Ok(Some(file)) => {
+                        let msg = VhostUserU64::new(0);
+                        self.main_sock
+                            .send_message(&reply_hdr, &msg, Some(&[file.as_raw_fd()]))?;
+                    }
+                    Err(_) => {
+                        let msg = VhostUserU64::new(0x101);
+                        self.main_sock.send_message(&reply_hdr, &msg, None)?;
+                    }
+                }
+            }
+            Ok(FrontendReq::CHECK_DEVICE_STATE) => {
+                let res = self.backend.check_device_state();
+                self.send_ack_message(&hdr, res)?;
```

Comment:
As far as I can see, `send_ack_message()` only sends a reply if `F_REPLY_ACK` has been negotiated. However, `CHECK_DEVICE_STATE` is defined to always send a reply, independently of `F_REPLY_ACK`. So I don’t think we should use `send_ack_message()` here (or if we do, I think it would need an additional parameter to force it to send the reply message).

Reply:
True

Reply:
(Looks good in the current version.)